### ISSUE TITLE
Jakarta EE 10: Support Jakarta JSON Binding 3.0

### ISSUE THREAD

**Issue by jhanders34** [OpenLiberty/open-liberty#18732]
## Description of the high level feature, including any external spec links:  
Implement the Jakarta JSON Binding 3.0 specification and provide jsonbContainer-3.0 functionality.

[JSON-B milestone 3.0](https://github.com/eclipse-ee4j/jsonb-api/milestone/2) includes:

- [x] 52 Allow `@JsonbNillable` at field and method level, and then deprecate nillable=true/false on `@JsonbProperty`
- [x] 71 Allow `@JsonbTypeDeserializer` and `@JsonbTypeAdapter` to be specified at parameter level
- [x] 121 Introduce `@JsonbRequired` to designate which `@JsonbCreator` fields are optional vs required.
- [x] 181 When deserializing a JsonValue typed attribute, a null value should be serialized as JsonValue.NULL rather than null
- [x] 147 (if added) Polymorphic serialization/deserialization 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/5f3qc2i72lhjsaxuxnnq9xxyngpkwhlh
- FTS: https://github.com/OpenLiberty/open-liberty/issues/21226
- Beta Blog Post (if applicable): https://github.com/openliberty/open-liberty/issues/21753
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/23288

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github. (requested)

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [x] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by jhanders34** [comment]
Notes for JSON-B from the UFO socialization that happened yesterday:

**Java APIs/SPIs - JSON-B 3.0 #52**
- What is the behavior with nillable=false and @JsonbNillable also specified on the same field?  Update the slide with what is the spec required behavior.

**Java APIs/SPIs - JSON-B 3.0 #121**
- Update slide to indicate that the @JsonRequired annotation can also be specified at the method and class level.  The example only shows parameter pattern.

**Migration Impact**
- Open a migration tools issue to be on the backlog to help customers when moving to new JSON versions.

**Comment by njr-11** [comment]
Slide 19 is updated to indicate the precedence for nillable.

While investigating JsonbRequired, I discovered that it had since been replaced. They decided not to introduce that annotation after all (it would have been new in 3.0) and instead opted for a simpler solution on JsonbConfig. I have updated slide 21 accordingly.

Migration tools issue opened: https://github.ibm.com/was-migration-tools/wamt/issues/5039

**Comment by tevans78** [comment]
Once a version of Yasson has been released with a fix for https://github.com/eclipse-ee4j/yasson/issues/557 there are a couple of tests that need to be re-enabled: https://github.com/OpenLiberty/open-liberty/issues/21051

**Comment by KyleAure** [comment]
Added `focalApproved:svt` on behalf of @hanczaryk who was having issues with GitHub

**Comment by KyleAure** [comment]
# Serviceability

## User Feature Overview (UFO) 

- Does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? 
     - Yes, the following scenarios have been identified and covered in the UFO. Prevent jsonb-3.0 and jsonbContainer-3.0 features from being enabled at the same time. Prevent jsonb-3.0 from being intermixed with another Jakarta EE version. Jsonb has deprecated `@JsonbProperty(nillable=true)` in favor of the annotation `@JsonbNillable`.  `@JsonbCreator` can now handle absent values during deserialization. 

- Have these issues been addressed in the implementation?
    - The first two scenarios are specific to Liberty and handles as part of our feature processing.
    - The last two scenarios have been handled by the Yasson implementation that we re-distribute to customers and have tested internally.

## Test and Demo
As part of the serviceability process, we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team). 

- What problem paths were tested and demonstrated? 
    - We performed regression testing to ensure our jsonb and jsonbContainer features worked correctly and interacted with the corresponding jsonp features in the same way as it did in Jakarta EE 9.  
    - Additionally, we automated the execution of the jsonb TCK which does more fine-grain unit testing.  We ran this TCK against the re-bundled Yasson library that we plan to distribute to customers to ensure compliance. 
- Who did you demo to? 
    - Presented features to developers on our team that were not involved in the development of the feature.  
    - We wrote and distributed a Skills Transfer document to L2.  We then demoed the feature and presented the skills transfer to our L2 support team.
- Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
    - There was a concern that there was no messaging for when a user-enabled jsonb and jsonbContainer features at the same time.  As a result, we added messaging to address this concern: https://github.com/OpenLiberty/open-liberty/pull/21489
    - There was a concern that prior serviceability reviews indicated that we would have an "avoiding trouble" document to show users what would happen in the jsonbContainer feature was enabled, but no BELL was configured on the server. As a result, we added documentation to address this concern: https://github.com/OpenLiberty/docs/pull/5590

## System Verification Testing

SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered. 

- Who conducted SVT tests for this feature? 
    - It was determined that no specific SVT testing was required for this feature. 
    - Brian Hanczaryk will be performing overall SVT testing for Jakarta EE 10 which will include jsonb.
- Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
   - During the process of implementing the jsonb TCK Brian Decker reached out with issues getting the TCK to run and pass alongside the platform TCK.  During that process, we discovered the issue was specific to the TCK and once resolved we were able to pass with minimal issues. 

## Service

- Which L2 / L3 queues will handle PMRs for this feature? 
    - L2: WAS L2 WIN
    - L3: WAS L3 JSON 
- Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.
    - Checked and verified that the Queue Contact Summary was updated and accurate 

## Metrics

- Does this feature add any new metrics or emit any new JSON events? 
    - No this feature does not add any new metrics or JSON Events

**Comment by donbourne** [comment]
see comments from https://github.com/OpenLiberty/open-liberty/issues/18733#issuecomment-1151566338

**Comment by chirp1** [comment]
From slack with Kyle and Kyle/David M, Kyle indicated that no documentation specific to this feature needs to be written. Approving.

**Comment by malincoln** [comment]
@samwatibm @steven1046 @jhanders34 can you provide focal approvals?  Thanks

**Comment by samwatibm** [comment]
1 .Are there translations?
2. If so, were they won't even be sent out until tomorrow, Friday Sep 23. (Unless they were sent in prior releases.)

**Comment by KyleAure** [comment]
@samwatibm 
1. Are there translations? 
  - It looks like the only translatable files that were updated via this feature were: 
  - https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.org.eclipse.yasson.3.0/resources/yasson-messages.properties
  - https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.websphere.appserver.features/visibility/public/jsonb-3.0/resources/l10n/io.openliberty.jsonb-3.0.properties
  - https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.websphere.appserver.features/visibility/public/jsonbContainer-3.0/resources/l10n/io.openliberty.jsonbContainer-3.0.properties
  - Which all look to be updated, but you may want to double-check?
